# A player has a 20% chance of hitting a pin in bowling. If they bowl 3 times, what’s the probability of hitting pins exatly 2 times?

The assumptions I am making is that there is a 20% chance of the player hitting a pin when the player does random moves. We are also assuming that it takes 700 steps to make three bowls as figured in testing.
The random variable I am measuring is the bowling ball hitting the pins.
One trial of the simulation is the player bowling 3 times.
A success outcome is the player hitting the pins exactly 2/3 times, failure is any other case.

In [14]:
!pip install gymnasium
!pip install "gymnasium[atari]"
!pip install ale-py
!pip install autorom[accept-rom-license]

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


In [15]:
import gymnasium as gym
import matplotlib.pyplot as plt
import ale_py

In [16]:

def runGame(i):
    print(f"Step: {i}", end="\r", flush=True)
    hits = 0

    # Create environment with no rendering
    env = gym.make('ALE/Bowling-v5', render_mode=None)
    observation, info = env.reset()
    
    total_reward = 0
    
    for step in range(700):
        # Take random action
        action = env.action_space.sample()
        observation, reward, terminated, truncated, info = env.step(action)
        total_reward += reward

        if reward > 0: #Reward increases when a pin is hit
            hits += 1

        # Stop if the game ends
        if terminated or truncated:
            observation, info = env.reset() 
            total_reward = 0

    env.close()
    return 1 if hits == 2 else 0

In [17]:
p = 0
for i in range(1000):
    p += runGame(i)

p/1000

0.448

The probability of hitting pins 2 out of 3 times empirically is 44.8% based on 1000 trials, the theorhetical probability is 9.6% based on a 20% chance of hitting a pin. The discrepancy between these values is likely due to my 20% assumption being too low.